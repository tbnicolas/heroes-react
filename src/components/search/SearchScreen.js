import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroeByName } from '../../selectors/getHeroeByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();

    console.log('Se ejecuto location');

    const { q= '' } = queryString.parse(location.search);

    const initialForm = {
        search:q
    }

    const [ {search}, handleInputChange ] = useForm(initialForm);

    
    const handleSearch = ( e ) => {
        e.preventDefault();
        history.push(`?q=${ search }`);
    }
    
    const heroesFiltered = useMemo(() => getHeroeByName(q), [q]);
    

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className ="col-5">
                    <h4> Search Form </h4>
                    <form onSubmit={ handleSearch }>
                        <input 
                            type="text"
                            name="search"
                            placeholder="Find your hero"
                            className="form-control"
                            value={ search }
                            onChange={ handleInputChange }
                        />
                        <button
                            type="submit"
                            className="btn mt-1 btn-outline-primary"
                            style={ { width: '100%' } }
                            >
                            Search...
                        </button>
                    </form>
                    
                </div>

                <div className ="col-7">
                    
                    <h4> Results </h4>
                    <hr/>

                    {
                        ( q==='')
                        &&
                        <div className="alert alert-info">
                            Search a hero
                        </div>
                    }

                    {
                        ( q !== '' && heroesFiltered.length === 0)
                        &&
                        <div className="alert alert-info">
                            There is no a hero with {q}
                        </div>
                    }

                    {
                        heroesFiltered.map((heroe)=>(
                            <HeroCard 
                                key={ heroe.id }
                                {...heroe}
                            />
                        ))

                    }

                </div>


            </div>
        </div>
    )
}
