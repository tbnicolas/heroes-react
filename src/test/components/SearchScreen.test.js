import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../components/search/SearchScreen";


describe('Pruebas en <SearchScreen />', () => {
   
    test('debe de mostrarse con valores por defecto', () => {
       
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a hero');
        
    });
    

    test('debe de mostrar a batman y el valor del query string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe mostrar un error si no se encuentra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-info').text().trim() ).toBe('There is no a hero with batman123');
    });


    test('debe de llamar el push del history', () => {
       
        const historyMock= {
            push: jest.fn()
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path="/search" component={
                    () => <SearchScreen history={historyMock}/>
                }/>
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target:{
                name: 'search',
                value:'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({preventDefault(){}})
        
        expect(historyMock.push).toHaveBeenCalledWith('?q=batman');
    });
    


    
    
});
