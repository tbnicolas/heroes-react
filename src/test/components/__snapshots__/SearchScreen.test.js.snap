// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a batman y el valor del query string 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-outline-primary"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="card card-space"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Ver m√°s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-outline-primary"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
